File: src/main/java/me/taubsie/dungeonhub/application/command/commands/CalcPriceCommand.java
package me.taubsie.dungeonhub.application.command.commands;

import me.taubsie.dungeonhub.application.command.Command;
import me.taubsie.dungeonhub.application.command.CommandParameters;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.CarryDifficultyConnection;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.CarryTierConnection;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.CarryTypeConnection;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.ServerConnection;
import me.taubsie.dungeonhub.application.enums.EmbedColor;
import me.taubsie.dungeonhub.application.exceptions.CommandExecutionException;
import me.taubsie.dungeonhub.application.exceptions.InvalidOptionException;
import me.taubsie.dungeonhub.application.service.ApplicationService;
import me.taubsie.dungeonhub.common.model.carry_difficulty.CarryDifficultyModel;
import me.taubsie.dungeonhub.common.model.carry_tier.CarryTierModel;
import org.javacord.api.entity.DiscordEntity;
import org.javacord.api.entity.channel.Categorizable;
import org.javacord.api.entity.channel.Channel;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.entity.server.Server;
import org.javacord.api.event.interaction.SlashCommandCreateEvent;
import org.javacord.api.interaction.SlashCommandOption;
import org.javacord.api.interaction.SlashCommandOptionBuilder;
import org.javacord.api.interaction.SlashCommandOptionType;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

@CommandParameters(name = "calc-price",
        description = "Calculate the price for some amount of carries.")
public class CalcPriceCommand extends Command {
    //TODO move to service
    public static long calculatePrice(CarryDifficultyModel carryDifficulty, long amount) {
        return calculatePricePerCarry(carryDifficulty, amount) * amount;
    }

    public static long calculatePricePerCarry(CarryDifficultyModel carryDifficulty, long amount) {
        Optional<Integer> bulkPrice = carryDifficulty.getBulkPrice();
        Optional<Integer> bulkAmount = carryDifficulty.getBulkAmount();

        if (bulkPrice.isPresent() && bulkAmount.isPresent() && bulkAmount.get() <= amount) {
            return bulkPrice.get();
        }

        return carryDifficulty.getPrice();
    }

    @Override
    protected void executeCommand(SlashCommandCreateEvent slashCommandCreateEvent) {
        Server server = getServer();

        long amount = getLongOption("amount");

        Optional<CarryTierModel> carryTier = slashCommandCreateEvent.getSlashCommandInteraction().getChannel()
                .flatMap(Channel::asCategorizable)
                .flatMap(Categorizable::getCategory)
                .map(DiscordEntity::getId)
                .flatMap(id -> ServerConnection.getInstance().getCarryTierFromCategory(server.getId(), id));

        try {
            Optional<CarryTierModel> previousCarryTier = carryTier;

            carryTier = CarryTypeConnection.getInstance(server.getId())
                    .getByIdentifier(getStringOption("carry-type"))
                    .flatMap(carryType -> CarryTierConnection.getInstance(carryType)
                            .getByIdentifier(getStringOption("carry-tier")))
                    .or(() -> previousCarryTier);
        }
        catch (CommandExecutionException commandExecutionException) {
            //ignored since the checking happens later
        }

        if (carryTier.isEmpty()) {
            throw new InvalidOptionException("carry-tier", "Please either use this in a carry ticket or supply a " +
                    "carry type and carry tier.");
        }

        Optional<CarryDifficultyModel> carryDifficulty = CarryDifficultyConnection.getInstance(carryTier.get())
                .getByIdentifier(getStringOption("carry-difficulty"));

        if (carryDifficulty.isEmpty()) {
            throw new InvalidOptionException("carry-difficulty");
        }

        long price = calculatePrice(carryDifficulty.get(), amount);
        long pricePerCarry = calculatePricePerCarry(carryDifficulty.get(), amount);

        if (price < 0) {
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "Something went wrong.. The calculated price (" + price + ") is negative?";
                }
            };
        }

        String priceText = price != 0
                ? ApplicationService.getInstance().makeNumberReadable(price) + " coins"
                : "Free";

        String pricePerCarryText = price != 0
                ? ApplicationService.getInstance().makeNumberReadable(pricePerCarry) + " coins"
                : "Free";

        EmbedBuilder embed = ApplicationService.getInstance()
                .getEmbed()
                .setColor(EmbedColor.INFORMATION.getColor())
                .setTitle("Carry-Price")
                .addInlineField("Type",
                        carryTier.get().getDisplayName() + " | " + carryDifficulty.get().getDisplayName())
                .addInlineField("Amount", String.valueOf(amount))
                .addInlineField("Price", priceText)
                .addInlineField("Price per Carry", pricePerCarryText);

        carryDifficulty.flatMap(CarryDifficultyModel::getThumbnailUrl).ifPresent(embed::setThumbnail);

        respond(embed);
    }

    @Override
    public List<SlashCommandOption> getSlashCommandOptions() {
        SlashCommandOption amountOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.LONG)
                .setName("amount")
                .setDescription("The amount of carries you want.")
                .setLongMaxValue(200)
                .setLongMinValue(1L)
                .setRequired(true)
                .build();

        return Arrays.asList(CarryTypeCommand.getCarryTypeOption(),
                CarryTierCommand.getCarryTierOption(),
                CarryDifficultyCommand.getCarryDifficultyOption(),
                amountOption);
    }
}

File: src/main/kotlin/me/taubsie/dungeonhub/application/commands/CalcPriceCommand.kt
package me.taubsie.dungeonhub.application.commands

import dev.kord.core.behavior.channel.asChannelOfOrNull
import dev.kord.core.entity.channel.CategorizableChannel
import dev.kordex.core.commands.Arguments
import dev.kordex.core.commands.converters.impl.long
import dev.kordex.core.commands.converters.impl.string
import dev.kordex.core.extensions.Extension
import dev.kordex.core.extensions.publicSlashCommand
import me.taubsie.dungeonhub.application.enums.EmbedColor
import me.taubsie.dungeonhub.application.exceptions.CommandExecutionException
import me.taubsie.dungeonhub.application.exceptions.InvalidOptionException
import me.taubsie.dungeonhub.application.loader.LoadExtension
import me.taubsie.dungeonhub.application.service.ApplicationService
import me.taubsie.dungeonhub.application.service.AutoCompletionService
import net.dungeonhub.connection.CarryDifficultyConnection
import net.dungeonhub.connection.CarryTierConnection
import net.dungeonhub.connection.CarryTypeConnection
import net.dungeonhub.connection.DiscordServerConnection

/**
 * Command to calculate the price for some amount of carries.
 * This command can be used in a carry ticket channel to automatically get the price for the current carry tier.
 * If used outside a carry ticket channel, the user has to supply the carry type and carry tier.
 * The command will then calculate the price for the given amount of carries.
 */
@LoadExtension
class CalcPriceCommand : Extension() {
    override val name = "calc-price-command"

    override suspend fun setup() {
        publicSlashCommand(::CalcPriceArguments) {
            name = "calc-price"
            description = "Calculate the price for some amount of carries."
            allowInDms = false

            action {
                respond {
                    var carryTier = channel.asChannelOfOrNull<CategorizableChannel>()
                        ?.categoryId
                        ?.let { id ->
                            DiscordServerConnection.getCarryTierFromCategory(
                                guild!!.id.value.toLong(),
                                id.value.toLong()
                            )
                        }

                    val previousCarryTier = carryTier

                    carryTier = CarryTypeConnection[guild!!.id.value.toLong()]
                        .getByIdentifier(arguments.carryType)
                        ?.let { carryType ->
                            CarryTierConnection[carryType]
                                .getByIdentifier(arguments.carryTier)
                                ?: previousCarryTier
                        }

                    if (carryTier == null) {
                        throw InvalidOptionException(
                            "carry-tier", "Please either use this in a carry ticket or supply a " +
                                    "carry type and carry tier."
                        )
                    }

                    val carryDifficulty = CarryDifficultyConnection[carryTier]
                        .getByIdentifier(arguments.carryDifficulty)
                        ?: throw InvalidOptionException("carry-difficulty")

                    val price = ApplicationService.calculatePrice(carryDifficulty, arguments.amount)
                    val pricePerCarry = ApplicationService.calculatePricePerCarry(carryDifficulty, arguments.amount)

                    if (price < 0) {
                        throw CommandExecutionException("Something went wrong.. The calculated price ($price) is negative?")
                    }

                    val priceText = if (price != 0L) "${ApplicationService.makeNumberReadable(price)} coins" else "Free"
                    val pricePerCarryText =
                        if (price != 0L) "${ApplicationService.makeNumberReadable(pricePerCarry)} coins" else "Free"

                    val embed = ApplicationService.embed
                    embed.color = EmbedColor.Information.color
                    embed.title = "Carry-Price"
                    embed.field("Type", true) {
                        carryTier.displayName + " | " + carryDifficulty.displayName
                    }
                    embed.field("Amount", true) { arguments.amount.toString() }
                    embed.field("Price", true) { priceText }
                    embed.field("Price per Carry", true) { pricePerCarryText }

                    carryDifficulty.thumbnailUrl?.let { embed.thumbnail { url = it } }

                    embeds = mutableListOf(embed)
                }
            }
        }
    }

    inner class CalcPriceArguments : Arguments() {
        val carryType by string {
            name = "carry-type"
            description = "The identifier of the carry type"
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryType
        }

        val carryTier by string {
            name = "carry-tier"
            description = "The identifier of the carry tier"
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryTier
        }

        val carryDifficulty by string {
            name = "carry-difficulty"
            description = "The identifier of the carry difficulty"
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryDifficulty
        }

        val amount by long {
            name = "amount"
            description = "The amount of carries you want."
            maxValue = 200
            minValue = 1
        }
    }
}